### DISCUSSION TITLE
compile api without certain modules

### DISCUSSION THREAD

**Discussion by legoraft** [FabricMC/fabric-api#1953] [Mod Dev Support]
Is it possible to compile your own version of the fabric API that excludes certain modules? I really want to use the API, but hate that it's so large because I only use 3 functions of it.

**Comment by OroArmor** [comment]
Check the [Readme](https://github.com/FabricMC/fabric#using-fabric-api-to-develop-mods)

**Reply by legoraft** [reply]
In which file do I need to put the modules I want to use, because that's not really specified. I also already tried commenting out modules I didn't need in settings.gradle, but that only outputted errors

**Reply by warjort** [reply]
You list the modules you want where your dependencies are defined, probably build.gradle

NOTE: The modules you are explicitly using might depend on other modules, 
many depend upon ````fabric-api-base```` which includes the shared base class for events.

e.g.  commands depends upon base
https://github.com/FabricMC/fabric/blob/1.18/fabric-command-api-v1/build.gradle

Also if you want to run minecraft in your test environment with other mods like REI, those other mods will likely only specify a dependency on the full fabric api. 
You will need to work out which fabric api modules they are using and add a ````testmodImplementation```` dependency for their additional usage. 
See the command example above for syntax.

**Reply by legoraft** [reply]
okay, so I need to set which modules I want to use in my mod config, I can't compile a seperate api version with a set of modules?

**Reply by warjort** [reply]
You can do whatever you like. It just depends upon your gradle skills.

To make your own cutdown api jar you would need to create your own parent project that replaces these:
https://github.com/FabricMC/fabric/blob/1.18/build.gradle
https://github.com/FabricMC/fabric/blob/1.18/settings.gradle

Or you can hack something together that takes the full jar and repackages it to remove stuff you don't want.

In both cases, you will be responsible for maintaining these artifacts going forward as the fabric api changes,
unlike the method described above where you just consume bits you need.

**Reply by legoraft** [reply]
okay so it's quite difficult, I was trying to make a smaller api jar for in my mods folder, but there isn't a simple list you can make in the current build.gradle or something?
EDIT: Found a very janky way to do it, by putting al unnecessary modules in the `devOnlyModules` function in `build.gradle`, that way the jar you finally use in the mods folder doesn't use the  modules you put in that function

**Reply by modmuss50** [reply]
I think its unclear what you asking for. I think what you want is to be able to take parts of the api to put into your mods folder?

The easyist way to do this is to download the full jar file, and then extract it as a zip file. In the META-INF/jars directory you will find all of the modules, you can take these out and put them into your mods folder as you want.

Do note that some of them depend on others, for example most modules depend on fabric-api-base.

**Reply by legoraft** [reply]
oh yeah that's also a smart way, what I was looking for was to compile the fabric api jar, but excluding modules I don't need for my mod, but still have just one api jar for in my mods folder if I use 3 modules for example

**Comment by warjort** [comment]
> You will need to work out which fabric api modules they are using and add a testmodImplementation dependency for their additional usage.

Or just create a test configuration that depends upon the full fabric api.
